implementations/rust/ockam/ockam_vault/src/vault.rs

```diff
@@ -36,86 +35,62 @@ use ockam_node::compat::asynchronous::RwLock;
 #[derive(Default, Clone)]
 pub struct Vault {
     pub(crate) data: VaultData,
+    pub(crate) storage: Option<Arc<dyn Storage>>,
 }
 
 #[derive(Default, Clone)]
 pub(crate) struct VaultData {
-    // TODO: make these private, and save automatically on modification.
-    pub(crate) entries: Arc<RwLock<BTreeMap<usize, VaultEntry>>>,
-    pub(crate) next_id: Arc<AtomicUsize>,
+    pub(crate) entries: Arc<RwLock<BTreeMap<KeyId, VaultEntry>>>,
 }
 
 impl Vault {
     /// Create a new SoftwareVault
-    pub fn new() -> Self {
+    pub fn new(storage: Option<Arc<dyn Storage>>) -> Self {
         Self {
             data: Default::default(),
+            storage,
         }
     }
 
-    /// Serialize a vault to bytes which may later be restored using
-    /// `SoftwareVault::deserialize`.
-    #[cfg(feature = "storage")]
-    pub async fn serialize(&self) -> Vec<u8> {
-        serialize(&self.data).await
-    }
-
-    /// Load a vault from the serialized format produced by `SoftwareVault::serialize`.
-    #[cfg(feature = "storage")]
-    #[tracing::instrument(err, skip_all)]
-    pub fn deserialize(data: &[u8]) -> ockam_core::Result<Self> {
-        let data = deserialize(data).map_err(|e| {
-            tracing::error!("Data loaded from vault failed to parse: {:?}", e);
-            crate::VaultError::StorageError
-        })?;
-        Ok(Self { data })
-    }
-
     /// Same as ```Vault::new()```
     pub fn create() -> Self {
-        Self::new()
+        Self::new(None)
     }
-}
 
-#[derive(Debug, Eq, PartialEq, Clone)]
-#[cfg_attr(feature = "storage", derive(serde::Serialize, serde::Deserialize))]
-pub(crate) struct VaultEntry {
-    key_id: Option<String>,
-    key_attributes: SecretAttributes,
-    key: SecretKey,
-}
+    pub(crate) async fn preload_from_storage(&self, key_id: &KeyId) -> Result<()> {
+        if self.data.entries.read().await.contains_key(key_id) {
+            return Ok(());
+        }
 
-impl VaultEntry {
-    pub fn key_id(&self) -> &Option<String> {
-        &self.key_id
-    }
-    pub fn key_attributes(&self) -> SecretAttributes {
-        self.key_attributes
-    }
-    pub fn key(&self) -> &SecretKey {
-        &self.key
-    }
-}
+        let storage = match &self.storage {
+            Some(s) => s,
+            None => return Ok(()),
+        };
 
-impl VaultEntry {
-    pub fn new(key_id: Option<String>, key_attributes: SecretAttributes, key: SecretKey) -> Self {
-        VaultEntry {
-            key_id,
-            key_attributes,
-            key,
-        }
+        let entry = match storage.load(key_id).await {
+            Ok(e) => e,
+            Err(_) => return Ok(()),
+        };
+
+        // Possible error due to concurrency, ignore it
+        let _ = self
+            .data
+            .entries
+            .write()
+            .await
+            .insert(key_id.clone(), entry);
```

Comment:
```suggestion
        self.data
            .entries
            .write()
            .await
            .insert(key_id.clone(), entry);
```
Maybe in the doc comment of `preload_from_storage` one could mention that concurrent invocations may overwrite each other's entry?

Reply:
We don't have a way to update entries in a Storage, so proper Storage trait implementation should either return the same entry each time for the given key_id, or return an error if there is no such Entry.
Therefore, it's fine if preload_from_storage was called multiple times concurrently for the same key_id. 1 of the calls succeeds, others fail silently and we should end up with correct state

Reply:
Other option could be acquiring write lock at the beginning of this function, and only then call the storage, but that should be less performant I imagine

Reply:
AFAICS only in-memory map entries are replaced and it may be just fine. Reading the comment made me think that this is considered an error but the map does not return an error we need to handle.